tests/specs/mcd/verification.k

```diff
@@ -23,23 +23,22 @@ module VERIFICATION-SYNTAX
-    // rule #WordPackAddrUInt8        (     ADDR , UINT8               ) => UINT8 *Int pow160 +Int ADDR                              requires #rangeAddress(ADDR) andBool #rangeUInt(8, UINT_8)
-
-    rule maxUInt48 &Int  UINT48_UINT48             => UINT48_1 requires UINT48_UINT48 ==Int #WordPackUInt48UInt48(UINT48_1, UINT48_2) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
-    rule maxUInt48 &Int (UINT48_UINT48 /Int pow48) => UINT48_2 requires UINT48_UINT48 ==Int #WordPackUInt48UInt48(UINT48_1, UINT48_2) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
+    rule ADDR_UINT48_UINT48 ==K #WordPackAddrUInt48UInt48(ADDR, UINT48_1, UINT48_2) => ADDR ==Int maxUInt160 &Int ADDR_UINT48_UINT48 andBool UINT48_1 ==Int maxUInt48 &Int (ADDR_UINT48_UINT48 /Int pow160) andBool UINT48_2 ==Int maxUInt48 &Int (ADDR_UINT48_UINT48 /Int pow208)
 
-    rule maxUInt160 &Int  ADDR_UINT48_UINT48              => ADDR     requires ADDR_UINT48_UINT48 ==Int #WordPackAddrUInt48UInt48(ADDR, UINT48_1, UINT48_2) andBool #rangeAddress(ADDR) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
-    rule maxUInt48  &Int (ADDR_UINT48_UINT48 /Int pow160) => UINT48_1 requires ADDR_UINT48_UINT48 ==Int #WordPackAddrUInt48UInt48(ADDR, UINT48_1, UINT48_2) andBool #rangeAddress(ADDR) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
-    rule maxUInt48  &Int (ADDR_UINT48_UINT48 /Int pow208) => UINT48_2 requires ADDR_UINT48_UINT48 ==Int #WordPackAddrUInt48UInt48(ADDR, UINT48_1, UINT48_2) andBool #rangeAddress(ADDR) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
+    rule UINT48_UINT48 ==K #WordPackUInt48UInt48(UINT48_1, UINT48_2) => UINT48_1 ==Int maxUInt48 &Int UINT48_UINT48 andBool UINT48_2 ==Int maxUInt48 &Int (UINT48_UINT48 /Int pow48)
+    rule (UINT48_2 *Int pow48) |Int (maskWordPackUInt48UInt48_1 &Int UINT48_UINT48) => #WordPackUInt48UInt48(maxUInt48 &Int UINT48_UINT48, UINT48_2                                 ) requires #rangeUInt(48, UINT48_2)
+    rule  UINT48_1             |Int (maskWordPackUInt48UInt48_2 &Int UINT48_UINT48) => #WordPackUInt48UInt48(UINT48_1                    , maxUInt48 &Int (UINT48_UINT48 /Int pow48)) requires #rangeUInt(48, UINT48_1)
 
-    rule maxUInt160 &Int  ADDR_UINT8              => ADDR  requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
-    rule maxUInt8   &Int (ADDR_UINT8 /Int pow160) => UINT8 requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
+    rule ADDR_UINT8 ==K #WordPackAddrUInt8(ADDR, UINT8) => ADDR ==Int maxUInt160 &Int ADDR_UINT8 andBool UINT8 ==Int maxUInt8 &Int (ADDR_UINT8 /Int pow160)
```

Comment:
So, essentially you specified how to unify two terms when one of them is a WordPack term.  My concern is that this could be applied even if not necessary, leading to blowing up the size of constraints and serious performance issue.  Also, you may want to have a different unification strategy for some cases, e.g., `#WordPackAddrUInt8(X, Y) ==K #WordPackAddrUInt8(Z, W)`.

You can just go with these for now, but keep in mind the potential issue.

Reply:
BTW, no range condition for `ADDR_UINT8` needed?

Reply:
The re-organization of the lemmas and added comments should help make it clearer. Basically `_ ==K #WordPack(_, _)` itself is a large predicate which unfolds into several smaller predicates, which makes it easier to reason about the individual values packed into the storage slot.